code/networking/src/validate_ip/ValidateIp.java

```diff
@@ -0,0 +1,35 @@
+import java.util.Arrays;
+import java.util.regex.Pattern;
+
+public class ValidateIp {
+    public static void main(String[] args) {
+        assert (isValid("172.16.254.1"));
+        assert (!isValid("172.16.254.01"));
+        assert (isValid("2001:0db8:85a3:0000:0000:8a2e:0370:7334"));
+        assert (isValid("2001:db8:85a3:0:0:8A2E:0370:7334"));
+        assert (!isValid("02001:0db8:85a3:0000:0000:8a2e:0370:7334"));
+    }
+
+    private static boolean isValid(String addr) {
+        if (addr.contains(".")) {
+            return isValidv4(addr);
+        }
+        if (addr.contains(":")) {
+            return isValidv6(addr);
+        }
+        return false;
+    }
+
+    private static boolean isValidv4(String addr) {
+        Pattern pattern = Pattern.compile("(?:(?:(?:[1-9][0-9]{0,2})|0)\\.){3}(?:(?:[1-9][0-9]{0,2})|0)");
```

Comment:
i tested this pattern against some regex test website, it seems something wrong with it, and i see we need to handle the case with zero shortcut ipv4 like : 127.1 or 127.0.1, i may suggest using something like this **(?:(?:25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9]?[0-9])\.){1,3}(?:25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9]?[0-9])**

Reply:
Are 127.1 and 127.0.1 legal?
If so, we might need to expand the readme too.

Reply:
yes, they are, i'm not fully aware of the nature of this project, but i think if want to make it a wide used library i guess i agree with you to expand the readme and handle such cases.